### ISSUE TITLE
Command to create rust-toolchain.toml in the current directory

### ISSUE THREAD

**Issue by ian-h-chamberlain** [rust-lang/rustup#2868]
Apologies if this has already been requested, but I couldn't seem to find any issues describing something like it, nor any feature already built-in to `rustup` to do this. If there is, please feel free to close this issue.

**Describe the problem you are trying to solve**

Frequently, when I create a new binary project, I want to pin a specific toolchain immediately after initializing the project, e.g.
```sh
cargo new --bin some-project
cat >some-project/rust-toolchain.toml <<EOF
[toolchain]
channel = "stable"
EOF
```

However, I frequently forget the precise format of the toolchain file, and I have to search in `rustup` documentation to find an example and copy-paste that into my project. It would be nice if there was a simple way to generate a file with reasonable defaults that I could then customize, similar to the generated `Cargo.toml` created by `cargo new` or `cargo init`.

**Describe the solution you'd like**

Add one more more new rustup commands to create a `rust-toolchain.toml` in the current directory. I think this could work similarly to `rustup override set` (take a toolchain argument), or just use the default / currently-set toolchain. The generated file could also include comments that briefly describe some of the fields and link to the documentation.

**Notes**

Example of one possible workflow:

```sh
cargo new --bin some-project
cd some-project
rustup init nightly . # or whatever the syntax of the command looks like
```
After which the contents of `some-project/rust-toolchain.toml` would look like this:
```toml
# See more keys and their definitions at https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file
[toolchain]
channel = "nightly"
components = []
targets = []
profile = "default"
```

Some possible command names:
 * `rustup new` (for parity with `cargo new`)
 * `rustup init` (possibly confusing with `rustup-init`)
 * `rustup generate-toolchain-file` or similar (very wordy)
 * `rustup pin`
 * A subcommand or flag to pass to `rustup override set` ?

**Comment by tombh** [comment]
I got here from googling "rustup create rust-toolchain file". Surprised there isn't a command for this. Ever since I lost some time with an outdated rustc version I've religiously copy and pasted the rust-toolchain file. So agreed, I'd personally expect `cargo init` to do it. But equally wouldn't be surprised if the responsibility fell on `rustup override set` or similar.

**Comment by SuperKenVery** [comment]
I would personally prefer that `cargo` locks the rust toolchain. Even if that's not OK, a `rustup lock` or something similar would be great. The current approach is quite some drawback on reproducibility, I think.

**Comment by flipchan** [comment]
any update on this? Shouldn't be super hard to do

**Comment by bn-c** [comment]
+1

**Comment by rami3l** [comment]
Based on the Zulip thread https://rust-lang.zulipchat.com/#narrow/channel/490103-t-rustup/topic/rust-toolchain.2Etoml.20init.20command/with/619666950, for this feature I propose the following implementation:

1. A new subcommand: `rustup pin`, which pins the active toolchain as a `rust-toolchain.toml` that is put in the CWD.
2. Take the active toolchain in either qualified or unqualified form (the latter by default, the former with an extra flag), and write it to `rust-toolchain.toml` along with the target/component list detected for that toolchain. The idea is that the user can call it like `rustup +nightly pin` for example.
3. If the file exists, rustup bails out with an error.

If @rust-lang/rustup is okay with that I'd love to claim this.

---

For further extension, the hope is to use the first parameter as the output file, which defaults to `rust-toolchain.toml`.

That being considered, this actually plays very well with my previous proposal https://github.com/rust-lang/rustup/issues/5025, in that if you want a specific file for CI scenarios, you will be able to do:

```console
> rustup +nightly pin rust-toolchain-ci.toml
```

To pin to a different TOML file, and when you want to use it, you will just need to run:

```console
> cargo +file=rust-toolchain-ci.toml --version
```

**Comment by integeresting** [comment]
As discussed in Zulip, I'd love to see this coming to rustup and I think this proposal plays nicely into the override hierarchy. Thanks to @rami3l for helping out here.

**Comment by djc** [comment]
`rustup pin` sounds good to me.

**Comment by ChrisDenton** [comment]
It feels to me like `override` and `pin` should ideally be under the same subcommand. They're closely related concepts albeit different enough that they are worth distinguishing to some degree.

**Comment by rami3l** [comment]
@ChrisDenton Noting the [five levels of rustup toolchain overrides](https://rust-lang.github.io/rustup/overrides.html):

> 1. The `+<toolchain>` override.
> 2. The `RUSTUP_TOOLCHAIN` environment variable.
> 3. The `rustup override set` override.
> 4. The `rust-toolchain.toml` override.
> 5. The default toolchain.

My understanding is that `rustup override` is only about point 3 for whatever [historical] reason, and we are talking about point 4 in this issue. Mixing the two in the same subcommand might get new users to think that `rustup override set` is also operating on point 4, which totally misses the point.

That said I'm totally for reorganizing this a bit, but that IMHO requires finding another subcommand other than `rustup override`.

**Comment by ChrisDenton** [comment]
4 and 3 seems closely related to.me. Unlike the others they are both specific to the directory. The main difference is that one is a shared config while the other is personal.

As I said, there does need to be a distinction between the two but I'm not sure if that warrants a top level command. The words "Pin" and "Override" just sound way too similar to me.

**Comment by djc** [comment]
I feel like `override` is kind of a hack of last resort and is named for actively replacing some other configuration, whereas pinning is more neutral and is commonly understood to improve reproducibility rather than undermining it.

**Comment by ChrisDenton** [comment]
Fair, enough. I have no strong objections to a new `Pin` command. I just worry about accumulating potentially confusable top-level commands.

**Comment by djc** [comment]
To be clear, that's also fair! Was thinking we could rebrand `override set` to be `pin --local` but I'm not sure it would be an overall improvement.

**Comment by rami3l** [comment]
> Fair, enough. I have no strong objections to a new `Pin` command. I just worry about accumulating potentially confusable top-level commands.

@djc @ChrisDenton If new toplevel subcommands are better avoided, would `rustup toolchain pin` be a better match? I agree that `rustup pin` is pretty short but normally the frequency of calling it should be pretty low...

**Comment by djc** [comment]
Yes, maybe it doesn't (yet) deserve the prime real estate of a top-level command.

**Comment by rami3l** [comment]
Okay, switching to `rustup toolchain pin` then. It's still in a draft state so the timing is quite convenient.
